js/src/momoa.d.ts

```diff
@@ -0,0 +1,113 @@
+
+//-----------------------------------------------------------------------------
+// Options
+//-----------------------------------------------------------------------------
+
+type MomoaMode = "json" | "jsonc";
+
+interface MomoaTokenizeOptions {
+    mode: MomoaMode;
+    ranges: boolean;
+}
+
+interface MomoaParseOptions {
+    mode: MomoaMode;
+    ranges: boolean;
+    tokens: boolean;
+}
+
+//-----------------------------------------------------------------------------
+// Nodes
+//-----------------------------------------------------------------------------
+
+export interface MomoaNode {
+    type: string;
+    loc?: MomoaLocationRange;
+    range?: MomoaRange;
+}
+
+/**
+ * The root node of a JSON document.
+ */
+export interface MomoaDocumentNode extends MomoaNode {
+    body: MomoaNode;
+}
+
+/**
+ * Strings, numbers, and booleans.
+ */
+interface MomoaValueNode<T> extends MomoaNode {
+    value: T;
+}
+
+export type MomoaStringNode = MomoaValueNode<string>;
+export type MomoaNumberNode = MomoaValueNode<number>;
+export type MomoaBooleanNode = MomoaValueNode<boolean>;
+
+/**
+ * An array element.
+ */
+export interface MomoaElementNode extends MomoaNode {
+    value: MomoaNode;
+}
+
+/**
+ * An array.
+ */
+export interface MomoaArrayNode extends MomoaNode {
+    elements: Array<MomoaElementNode>;
+}
+
+/**
+ * An object member.
+ */
+export interface MomoaMemberNode extends MomoaNode {
+    name: MomoaStringNode;
+    value: MomoaNode;
+}
+
+/**
+ * An object.
+ */
+export interface MomoaObjectNode extends MomoaNode {
+    members: Array<MomoaMemberNode>;
+}
+
+type MomoaAnyNode = MomoaDocumentNode | MomoaArrayNode | MomoaBooleanNode |
+    MomoaElementNode | MomoaMemberNode | MomoaBooleanNode | MomoaStringNode |
+    MomoaNumberNode | MomoaNode;
+
+/**
+ * Additional information about the node.
+ */
+export interface MomoaNodeParts {
+    loc?: MomoaLocationRange;
+    range?: MomoaRange;
```

Comment:
all-optional interfaces are dangerous because anything is assignable to them.
although looking at the usage, this looks to be the way they are used, in fact.

Reply:
Understood. This is only used internally so I think it's okay.